Fix BackendConfig.securityPolicy is not removed after config updates

songrx1997 opened the pull request:
Fix [issue](https://github.com/kubernetes/ingress-gce/issues/1706) #1706 BackendConfig.securityPolicy is not removed after config updates

songrx1997 commented:
/assign @swetharepakula

swetharepakula commented:
/ok-to-test

swetharepakula commented:
Thanks for the change Ruixian! My main comment is that we want to still be able to exit early if we see that no change needs to be done for security policy. Can you also add a unit test for this change?

songrx1997 commented:
Hi @swetharepakula thanks for you comment.

> Thanks for the change Ruixian! My main comment is that we want to still be able to exit early if we see that no change needs to be done for security policy. 
For this, we need to know the desired security policy and current security policy. If they are different, we need to change the policy. Otherwise we do nothing and exit. So I think only after [getting both the desired and current policy](https://github.com/kubernetes/ingress-gce/pull/1749/commits/a5bbb02dc540961ab69644ff95c93e2ae43db789#diff-694a3074761a6ab43364827c08bb7533a5532940ce0c8a11c418f4ed266d9ce2R51) can we know whether changes need to be done. 

> Can you also add a unit test for this change?
Will do this, thanks for reminding!

songrx1997 commented:
Add a new [commit](https://github.com/kubernetes/ingress-gce/pull/1749/commits/143afcc0f94e36f0942e119c1aa511845365e59a) to update the test cases.

Added a new test case `remove-policy-2` for the situation:
Current Backend Service with securityPolicy &&  BackendConfig without securityPolicy ->  need to make a call to remove the policy. (This is the case in #1706 )

Removed `no-specified-policy`. We should make a call and remove the policy in this situation. So the case is no longer applicable.

songrx1997 commented:
/ok-to-test

songrx1997 commented:
/retitle

swetharepakula commented on pkg/backends/features/securitypolicy_test.go:
```diff
@@ -78,7 +78,7 @@ func TestEnsureSecurityPolicy(t *testing.T) {
 		},
 		{
-			desc: "remove-policy",
+			desc: "remove-policy-1",
```
nit: maybe rename to `change-security-policy`

songrx1997 replied:
Thanks @swetharepakula , this case would unset the security policy since the name field in the config is an empty string. I think `unset-policy-empty-policy-string` would be more clear.

swetharepakula commented on pkg/backends/features/securitypolicy_test.go:
```diff
@@ -92,6 +92,15 @@ func TestEnsureSecurityPolicy(t *testing.T) {
 			expectSetCall: true,
 		},
+		{
+			desc: "remove-policy-2",
```
nit: rename to `remove-policy` or `unset-policy`

songrx1997 replied:
Thanks @swetharepakula , this case would unset the security policy but it is different than the previous one. This one have no security policy field in the config. So I think `unset-policy-no-policy` would be more clear.

swetharepakula commented on pkg/backends/features/securitypolicy_test.go:
```diff
@@ -111,15 +120,6 @@ func TestEnsureSecurityPolicy(t *testing.T) {
-			desc: "no-specified-policy",
-			currentBackendService: &composite.BackendService{
-				Scope:          meta.Global,
-				SecurityPolicy: "https://www.googleapis.com/compute/projects/test-project/global/securityPolicies/policy-1",
```
modify this test so that the SecurityPolicy is not set in the currentBackendService

songrx1997 replied:
Thanks @swetharepakula , this is the `unset-policy-no-policy` test. Since it is similar to `unset-policy-empty-policy-string`, I put them together to make it clear.

swetharepakula replied:
I think "regional backend service with no specified security policy" covers the case I was thinking of.

swetharepakula reviewed (commented):
Few nits left. This is almost ready to merge. Thanks!

swetharepakula commented on pkg/backends/features/securitypolicy.go:
```diff
@@ -31,20 +31,23 @@ import (
 		return fmt.Errorf("cloud armor security policies not supported for %s backend service %s", be.Scope, be.Name)
 	}
 
+	var desiredPolicyName string
```
@spencerhance, is there any problem because we are checking the scope before checking if security policy is nil? I am wondering if there are some cases where we will accidentally return an error but previously we would have moved on because security policy was nil?

spencerhance replied:
Hm I think it should be okay because the policies would have to be different as well as the scope being regional.  Since you can't set the security policy on a regional BackendService I think we should be okay.

swetharepakula replied:
My comment isn't applicable anymore. The previous version checked the scope before checking if there was a change to the security policy. I think this version of the code doesn't run into the issue I mentioned above. Thanks for reviewing Spencer!

spencerhance commented on pkg/backends/features/securitypolicy.go:
```diff
@@ -31,24 +31,27 @@ import (
+	if sp.BackendConfig.Spec.SecurityPolicy != nil {
+		desiredPolicyName = sp.BackendConfig.Spec.SecurityPolicy.Name
+	} else {
+		desiredPolicyName = ""
```
desiredPolicyName should be an empty string by default right?

songrx1997 replied:
Yes it should be an empty string.

spencerhance reviewed (approved):
One possible scenario is that there is some customer who got stuck in this state accidentally and will be suprised when their policy is removed.  That being said, it's probably a very rare occurrence given how important securityPolicy is.

songrx1997 commented:
> One possible scenario is that there is some customer who got stuck in this state accidentally and will be suprised when their policy is removed. That being said, it's probably a very rare occurrence given how important securityPolicy is.

Thanks @spencerhance Do you mean the situation that customers forget to put SecurityPolicy in the backendConfig and accidentally unset their security policy?

swetharepakula reviewed (commented):
Thanks for the change Ruixian!

/lgtm
/approve
